javascript-es2015/src/chapter3/ch3-q6.js

```diff
@@ -0,0 +1,55 @@
+'use strict';
+
+/**
+ * Uses two different queues one for dogs and one for cats. Each entry is
+ * assigned a unique identifier which allows dequeueAny to determine which of
+ * the two queues to dequeue an item from.
+ *
+ * N = number of animals
+ * Time: enqueue O(1), dequeue O(1), dequeueAny O(1)
+ * Additional space: enqueue O(N), dequeue O(1), dequeueAny O(1)
+ * Additional space required to hold unique id per animal.
+ */
+export class AnimalShelter {
+  constructor() {
+    this._dogs = [];
+    this._cats = [];
+    this._id = 0;
+  }
+
```

Comment:
enqueueAny is missing. Whether you should use a class Dog and Cat and check instanceof (if you like es6 classes :-) or the object passed to enqueueAny should have a type that you check to call then enqueueDog or enqueueCat.